[Tan Ming Wei] iP

malwaregarry opened the pull request:
## Duke 😎
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is a CLI task list manager. It's
* text-based
* made with ~~Python~~ Java 
### Features
1. Task management
2. *and more...*
### What's planned
- [ ] Graphical UI
- [ ] Code refactoring

Download it [here](https://github.com/malwaregarry/ip/releases) for **FREE**!
 ___
Here's the ```main``` method:

```java
public static void main(String[] args) {
    new Duke(STORAGE_PATH).run();
}
```

Commit:
Add Gradle support

malwaregarry committed:
implement greet, echo, exit

malwaregarry committed:
Add 'add, list' to duke

malwaregarry committed:
Implement marking and unmarking tasks as done

malwaregarry committed:
Add Deadline, Event and Todo subclasses

malwaregarry committed:
Read command from first word in line instead of entire line

malwaregarry committed:
Add ToDos, Events, Deadlines

malwaregarry committed:
Test application with automated text ui testing

malwaregarry committed:
Create DukeException class

For handling exceptions related to Duke

malwaregarry committed:
Handle errors from Duke class

Errors handled:
* Unknown command
* Empty descriptions for tasks
* Improper formatting of inputs for Deadline and Event creation

malwaregarry committed:
Add option to delete tasks

malwaregarry committed:
Add UI tests for delete

malwaregarry committed:
Change listener to use Command enums

malwaregarry committed:
Add saving of tasks to storage file

malwaregarry committed:
Change Event & Deadline to store date using LocalDate

malwaregarry committed:
Merge branch 'branch-Level-7'

malwaregarry committed:
Fix test for Date changes

malwaregarry committed:
Add updateLine method in Storage for mark/unmark

malwaregarry committed:
Refactor Commands into their own type

malwaregarry committed:
Refactor program to make use of more OOP

malwaregarry committed:
Rearrange imports to meet coding standards

malwaregarry committed:
Divide classes into packages

malwaregarry committed:
Add public modifier for some methods

malwaregarry committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

malwaregarry committed:
Add gradle support

malwaregarry committed:
Add JUnit tests for Command and Ui

malwaregarry committed:
Add JavaDoc comments for public methods and clases

malwaregarry committed:
Update code to meet coding standards

malwaregarry committed:
Add FindCommand

malwaregarry committed:
Merge branch 'branch-A-CodingStandard'

malwaregarry committed:
Merge branch 'branch-Level-9'

malwaregarry committed:
Create branch for iP progress

malwaregarry committed:
Implement checkstyle

malwaregarry committed:
Fix DeleteCommand not deleting from Storage

malwaregarry committed:
Capitalize params in JavaDoc comments

malwaregarry committed:
Remove trailing comma in vscode settings.json

malwaregarry committed:
Update Command and Ui class for future GUI support

malwaregarry committed:
Update Duke to be use a GUI with JavaFX

malwaregarry committed:
Merge branch 'branch-A-CheckStyle'

malwaregarry committed:
Merge branch 'branch-Level-10'

malwaregarry committed:
Merge branch 'branch-Level-10'

malwaregarry committed:
Merge branch 'branch-Level-10'

malwaregarry committed:
Update reference in README

malwaregarry committed:
Merge branch 'branch-Level-10'

fannyjian commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,283 @@
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+
```
I like how you listed out all your imports to follow the coding standards!

fannyjian commented on src/main/java/duke/StorageInterface.java:
```diff
@@ -0,0 +1,44 @@
+public interface StorageInterface {
+
+    /**
+     * Read storage file and return data in file as a list of Tasks.
```
I think this line should start with "Reads" instead of "Read"!

fannyjian commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,283 @@
+ * Class used to parse user inputs.
+ */
+public abstract class Parser {
+    private static final String BYE = "bye";
+    private static final String LIST = "list";
+    private static final String TODO = "todo";
+    private static final String DEADLINE = "deadline";
+    private static final String EVENT = "event";
+    private static final String MARK = "mark";
+    private static final String UNMARK = "unmark";
+    private static final String DELETE = "delete";
+    private static final String FIND = "find";
+    private static final String DEADLINE_INDICATOR_PATTERN = "\\s*/by\\s*";
+    private static final String EVENT_INDICATOR_PATTERN = "\\s*/at\\s*";
```
Good naming of constants!

rjkoh replied:
i like how you check the commands using constants, and follow the naming conventions. the indicator patterns are also well done in that they are complicated enough that users will not accidentally input the same pattern due to typos!

fannyjian commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,283 @@
+        Optional<String> commandArguments = commandString.getValue();
+
+        Command command;
+        switch (commandIndicator) {
+        case BYE:
+            command = parseBye();
+            break;
+        case LIST:
+            command = parseList();
+            break;
+        case TODO:
+            command = parseTodo(getCommandArgument(commandArguments));
+            break;
+        case DEADLINE:
+            command = parseDeadline(getCommandArgument(commandArguments));
+            break;
+        case EVENT:
+            command = parseEvent(getCommandArgument(commandArguments));
+            break;
+        case MARK:
+            command = parseMark(getCommandArgument(commandArguments));
+            break;
+        case UNMARK:
+            command = parseUnmark(getCommandArgument(commandArguments));
+            break;
+        case DELETE:
+            command = parseDelete(getCommandArgument(commandArguments));
+            break;
+        case FIND:
+            command = parseFind(getCommandArgument(commandArguments));
+            break;
+        default:
```
Correct avoidance of indentation for case clauses!

fannyjian reviewed (commented):
Overall, your code complies with the coding standards, except for a minor mistake pointed out!

malwaregarry committed:
Fix JavaFX version in FXML and reference code used

rjkoh commented on src/main/java/duke/StorageStub.java:
```diff
@@ -0,0 +1,44 @@
+    public List<Task> getTasks() {
+        return this.tasks;
+    }
+}
```
I like the use of stubs for testing, since your Storage class is detailed!

rjkoh commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,33 @@
+        Command.taskList.add(this.task);
+        Command.storage.save(this.task);
+        return Ui.getAddTaskMessage(this.task, Command.taskList.size());
+    }
```
I like how the commands are in their own classes and have their own execution implemented in the class itself (instead of being run outside)

rjkoh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,283 @@
+    static Command parse(String input) throws DukeException {
+        Map.Entry<String, Optional<String>> commandString = getCommandAndArguments(input);
+        String commandIndicator = commandString.getKey();
+        Optional<String> commandArguments = commandString.getValue();
```
I like your implementation of parse here as it uses Optional to account for missing or bad inputs. However, it takes some analysis due to the other implemented methods used. 

Perhaps you could consider adding some comments to make it easier to follow as I did have some trouble doing so.

rjkoh reviewed (commented):
I enjoyed looking at your code and learnt quite a bit from it! I particularly like how you broke the commands down, and implemented an interface and stub for storage. Your implementation is very detailed and accounts for many potential source of errors!

One suggestion would simply be to add some comments to make your code easier to follow since it is more complicated than most implementations! Well done though!

malwaregarry committed:
Fix DialogBox being truncated

malwaregarry committed:
Change entry point for application in build.gradle

malwaregarry committed:
Add assertions

Assertions can help identify bugs in the code.

Having assertions for pre-conditions, post-conditions
and invariant can identify bugs in regressions.

Let's add more assertions for future methods that require them.:w
:q
:q
:wq

malwaregarry committed:
Improve code quality

There are many magic literals and long methods in the code.

Using constants for magic literals make the code more readable.
Refactoring functions to adhere to SLAP will also make the code
more readable.

Let's
* stop using magic literals
* adhere to SLAP

malwaregarry committed:
Fix AddCommand test

malwaregarry committed:
Create Github Actions workflow file

malwaregarry committed:
Merge branch 'master' of github.com:malwaregarry/ip

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Merge branch 'master' of github.com:malwaregarry/ip

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Implement task description update

There is currently no way to update tasks. Users might
want to change the description of tasks.

An update command allows user to update task description.

This allows user to change task description.

malwaregarry committed:
Merge branch 'master' into branch-A-Assertions

malwaregarry committed:
Merge pull request #1 from malwaregarry/branch-A-Assertions

Add assertions

malwaregarry committed:
Merge branch 'master' into branch-A-CodeQuality

malwaregarry committed:
Fix assertions and missing semicolon

malwaregarry committed:
Change line separator to be based on System

malwaregarry committed:
Merge pull request #2 from malwaregarry/branch-A-CodeQuality

Improve code quality

malwaregarry committed:
Remove update message magic literal

malwaregarry committed:
Update user guide

malwaregarry committed:
Update user guide with example output

malwaregarry committed:
Add Ui screenshot

malwaregarry committed:
Add Bye command to user guide

malwaregarry committed:
Add config file for Github Pages

malwaregarry committed:
Add config to user guide

malwaregarry committed:
Update user guide

malwaregarry committed:
Change user guide theme to 'tactile'

malwaregarry committed:
Update CI workflow file

malwaregarry committed:
Update user guide

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Change boolean  to align with coding standards

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Update gradle.yml

malwaregarry committed:
Merge branch 'master' of github.com:malwaregarry/ip

malwaregarry committed:
Add screenshot to user guide and fix formatting

malwaregarry committed:
Fix window not closing with 'bye' command

malwaregarry committed:
Fix javadocc formatting error

malwaregarry committed:
Update gitignore

malwaregarry committed:
Update javafx version
